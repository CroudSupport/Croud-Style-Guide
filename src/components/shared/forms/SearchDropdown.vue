<template>
    <semantic-form-dropdown value="" :options="[]" :settings="computedSettings" :placeholder="placeholder" v-bind="$attrs" search/>
</template>

<script>
    import _ from 'lodash'

    /**
    * A search dropdown based on the vue-semantic form dropdown
    *
    * @example ./croud-search-dropdown.md
    */
    export default {
        name: 'croud-search-dropdown',

        props: {
           /**
            * Placeholder text
            */
            placeholder: {
                type: String,
                default: 'Search...',
            },

           /**
            * request part of the url
            */
            url: {
                type: String,
                required: true,
            },

           /**
            * Optional v-model alias
            */
            value: {
                required: false,
                type: [String, Number],
            },

           /**
            * Method to manipulate results back from the API
            */
            transform: {
                type: Function,
                default(result) {
                    return result
                },
            },

            /**
             * Semantic Dropdown settings
             */
            settings: {
                type: Object,
                default() {
                    return {}
                },
            },
        },

        computed: {
            model: {
                get() {
                    return this.value
                },
                set(value) {
                    this.$emit('input', value)
                },
            },

            computedUrl() {
                if (this.url.match(/:\/\//)) return this.url

                return `//${gateway_url}/${this.url}`
            },

            computedSettings() {
                const $this = this
                return _.defaultsDeep(this.settings, {
                    apiSettings: {
                        url: this.computedUrl,
                        beforeXHR(xhr) {
                            xhr.setRequestHeader('authorization', `Bearer ${localStorage.getItem('jwt')}`)
                            return xhr
                        },
                        onResponse(response) {
                            response.results = response.data.map($this.transform)
                            return response
                        },
                    },
                    fields: {
                        results: 'data',
                        title: 'name',
                        value: 'id',
                    },
                    selectOnKeydown: false,
                    searchDelay: 200,
                    onChange: (...args) => {
                        this.$emit('selected', ...args)
                        this.model = args[0]
                    },
                })
            },
        },
    }
</script>
