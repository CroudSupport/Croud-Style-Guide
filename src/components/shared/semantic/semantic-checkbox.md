### semantic-checkbox
For our checkboxes, we use the semantic-checkbox component from the [croud-vue-semantic](https://github.com/CroudSupport/vue-semantic) library. Below are a few examples of semantic-checkbox in action. Click [here](http://croudsupport.github.io/vue-semantic/#semantic-checkbox) for full component docs.

### Basic usage

    <semantic-checkbox v-model="checkbox" label="🍕 Pizza" />

### Styling
Pass through the **type** prop to change the style of checkbo

    <div class="ui form">
        <div class="ui field">
            <semantic-checkbox v-model="checkbox" label="Checkbox" />
        </div>
        <div class="ui field">
            <semantic-checkbox v-model="checkbox" label="Slider" type="slider"/>
        </div>
        <div class="ui field">
            <semantic-checkbox v-model="checkbox" label="Toggle" type="toggle" />
        </div>
    </div>

### Disabled

    <semantic-checkbox label="🍕 Pizza" type="toggle" :disabled="true"/>
