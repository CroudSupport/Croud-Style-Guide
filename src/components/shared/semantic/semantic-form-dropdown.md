### semantic-form-dropdown
For our form dropdowns, we use the semantic-form-dropdown component from the [croud-vue-semantic](https://github.com/croudtech/vue-semantic) library. Below are a few examples of semantic-form-dropdown in action. Click [here](http://croudtech.github.io/vue-semantic/#semantic-form-dropdown) for full component docs.

### Basic usage

    <semantic-form-dropdown v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]"></semantic-form-dropdown>

### Styling
You can use the props to change the styling of the dropdown

    <div class="ui form">
        <div class="field">
            <label>Fluid</label>
            <semantic-form-dropdown fluid v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" />
        </div>

        <div class="three fields">
            <div class="field">
                <label>Search</label>
                <semantic-form-dropdown search v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" />
            </div>
            <div class="field">
                <label>Multiple</label>
                <semantic-form-dropdown multiple v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" />
            </div>
            <div class="field">
                <label>Disabled</label>
                <semantic-form-dropdown :disabled="true" v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" />
            </div>
        </div>

        <div class="inline field">
            <label>Inline</label>
            <semantic-form-dropdown :inline="true" v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" />
        </div>
    </div>

### Using slots
The **default** slot allows you to add custom options to the top of the dropdown

    <semantic-form-dropdown v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]">
        <div class="item" v-data="3">
            <b>Baz</b>
        </div>
    </semantic-form-dropdown>

### Handle Events
This component emits the **dropdown-selected** event when an item is selected.

    <semantic-form-dropdown :value="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" @dropdown-selected="$toasted.success('TaDa!!!')"></semantic-form-dropdown>
