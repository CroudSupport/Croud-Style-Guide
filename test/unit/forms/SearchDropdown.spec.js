import Vue from 'vue'
import VueSemantic from 'croud-vue-semantic'
import SearchDropdown from '../../../src/components/shared/forms/SearchDropdown'
import '../../../semantic/dist/semantic'

Vue.use(VueSemantic)

const Constructor = Vue.extend(SearchDropdown)

const vm = new Constructor({
    propsData: {
        url: 'api/users?filter={query}',
    },
}).$mount()

describe('Search Dropdown', () => {
    it('should match the snapshot', () => {
        expect(vm.$el).toMatchSnapshot()
    })
})
