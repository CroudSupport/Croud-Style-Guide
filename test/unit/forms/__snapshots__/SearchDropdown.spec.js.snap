// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Dropdown should match the snapshot 1`] = `
<div
  class="ui selection dropdown search"
>
  <input
    type="hidden"
    name=""
  />
  <input
    class="search"
    autocomplete="off"
    tabindex="0"
  />
  <div
    class="default text"
  >
    Search...
  </div>
  <i
    class="dropdown icon"
  />
  <div
    class="menu dropdown_menu"
    tabindex="-1"
  >
    
  </div>
</div>
`;
